pkg/kubefed2/unjoin.go

```diff
@@ -0,0 +1,409 @@
+			return err
+		}
+
+		glog.V(2).Infof("Deleted federation namespace: %s from unjoin cluster: %s",
+			federationNamespace, unjoiningClusterName)
+	}
+
+	return nil
+}
+
+// performPreflightChecksForUnjoin checks that the host and unjoining clusters are in
+// a consistent state.
+func performPreflightChecksForUnjoin(clusterClientset client.Interface, name, host,
+	federationNamespace string) error {
+	// Make sure there is a existing service account in the unjoining cluster.
+	saName := util.ClusterServiceAccountName(name, host)
+	sa, err := clusterClientset.CoreV1().ServiceAccounts(federationNamespace).Get(saName,
+		metav1.GetOptions{})
+
+	if errors.IsNotFound(err) {
+		return fmt.Errorf("service account does not exist in unjoining cluster")
+	} else if err != nil {
+		return err
+	} else if sa != nil {
+		return nil
+	}
+
+	return nil
+}
+
+// removeFromClusterRegistry handles removing the cluster from the cluster registry and
+// reports progress.
+func removeFromClusterRegistry(hostConfig *rest.Config, host, unjoiningClusterName string,
+	dryRun bool) error {
+
+	// Get the cluster registry clientset using the host cluster config.
+	crClientset, err := util.ClusterRegistryClientset(hostConfig)
+	if err != nil {
+		glog.V(2).Infof("Failed to get cluster registry clientset: %v", err)
+		return err
+	}
+
+	glog.V(2).Infof("Removing cluster: %s from the cluster registry.", unjoiningClusterName)
+
+	err = unRegisterCluster(crClientset, host, unjoiningClusterName, dryRun)
+	if err != nil {
+		glog.V(2).Infof("Could not remove cluster from the cluster registry: %v", err)
+		return err
+	}
+
+	glog.V(2).Infof("Removed cluster: %s from the cluster registry.", unjoiningClusterName)
+	return nil
+}
+
+// unRegisterCluster removes a cluster from the cluster registry.
+func unRegisterCluster(crClientset *crclient.Clientset, host, unjoiningClusterName string,
+	dryRun bool) error {
+	if dryRun {
+		return nil
+	}
+
+	err := crClientset.ClusterregistryV1alpha1().Clusters(controllerutil.MulticlusterPublicNamespace).Delete(unjoiningClusterName,
+		&metav1.DeleteOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// deleteFederatedClusterAndSecret deletes a federated cluster resource that associates
+// the cluster and secret.
+func deleteFederatedClusterAndSecret(hostClientset client.Interface, fedClientset *fedclient.Clientset,
+	federationNamespace, unjoiningClusterName string, dryRun bool) (*fedv1a1.FederatedCluster, error) {
+	if dryRun {
+		return nil, nil
+	}
+
+	fedCluster, err := fedClientset.CoreV1alpha1().FederatedClusters(federationNamespace).Get(
+		unjoiningClusterName, metav1.GetOptions{})
+	if err != nil {
+		return fedCluster, err
+	}
+
+	err = fedClientset.CoreV1alpha1().FederatedClusters(federationNamespace).Delete(
+		unjoiningClusterName, &metav1.DeleteOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	err = hostClientset.CoreV1().Secrets(federationNamespace).Delete(fedCluster.Spec.SecretRef.Name,
```

Comment:
If the federated cluster is removed prior to the auth secret, it will not be possible to retry removal.  This suggests a reversal of the order of these operations.

Reply:
@font has different opinion for this https://github.com/kubernetes-sigs/federation-v2/pull/137#discussion_r202205702 , but after think more, yes, I will move reverse the order

Reply:
See my comment above. The credentials from the unjoining cluster context in the kubeconfig is used to access the joined cluster for removing resources, not the secret stored in the host cluster. 

My concern is that removal of resources should be in reverse order from their creation in the join because there could be unnecessary transient errors if one were to remove credential resources before effectively removing the federated cluster. Removing the federated cluster first, signals that this cluster is no longer available (albeit without checking for its successful removal) before attempting to remove any of its credential resources.

Reply:
Hmm, I do not have strong options for this. Here I was deleting `secret` and in `performPreflightChecksForUnjoin`, we are using `serviceAccount` to check if the unjoin cluster exist or not, yes, they are different.